[compiler-rt] Add `DumpAllRegisters` impl

chestnykh opened the pull request:
- Add implementation for x86_64 and linux
- Add test

The output is like

==XXYYZZ==Register values:
rax = 0x...  rbx = 0x...  rcx = 0x...  rdx = 0x...
rdi = 0x...  rsi = 0x...  rbp = 0x...  rsp = 0x...
 r8 = 0x...   r9 = 0x...  r10 = 0x...  r11 = 0x...
r12 = 0x...  r13 = 0x...  r14 = 0x...  r15 = 0x...

chestnykh committed:
[compiler-rt] Add `DumpAllRegisters` impl

- Add implementation for x86_64 and linux
- Add test

The output is like
`
==XXYYZZ==Register values:
rax = 0x...  rbx = 0x...  rcx = 0x...  rdx = 0x...
rdi = 0x...  rsi = 0x...  rbp = 0x...  rsp = 0x...
 r8 = 0x...   r9 = 0x...  r10 = 0x...  r11 = 0x...
r12 = 0x...  r13 = 0x...  r14 = 0x...  r15 = 0x...
`

chestnykh committed:
[compiler-rt] Suppress compiler warnings

Fox example gcc warns about `RegNumToRegName` is unused
and about `ucontext` is unused

chestnykh commented:
CC: @alexander-shaposhnikov  @vitalybuka

vitalybuka commented on compiler-rt/lib/sanitizer_common/sanitizer_linux.cpp:
```diff
@@ -2118,8 +2118,91 @@ bool SignalContext::IsTrueFaultingAddress() const {
+#  if SANITIZER_LINUX
+  ucontext_t *ucontext = (ucontext_t *)context;
+#    define DUMPREG64(r)                             \
+      Printf("%s = 0x%016llx  ", RegNumToRegName(r), \
```
mac version uses preprocessor for `#r`
if you replace that with **RegNumToRegName**, we don't need macro any more
lambda or function will be nicer.

vitalybuka replied:
Also can we just move  SignalContext::DumpAllRegisters from MAC ?

chestnykh replied:
Yeah, i thought about it.
copy-paste is not possible because of some difference in ucontext implementation for linux and mac.
And what should we write after `REQUIRES:` in the test if there will be an implementation for arm-linux, i386-linux, x86_64-linux and aarch64-linux? Can we require only `linux`?

chestnykh replied:
The differences in ucontext definition is the reason why i made RegNumToRegName

chestnykh replied:
I think that i will handle your suggestion about removing macroses from the `DumpAllRegisters` code for x86-64 and i386 but arm and aarch64 implementations will be done later because there is a need to change MAC's code and i don't have arm/aarch64 environment to test it :(

vitalybuka commented on compiler-rt/test/sanitizer_common/TestCases/Posix/dump_registers.cpp:
```diff
@@ -1,16 +1,20 @@
-// FIXME: Implement.
-// XFAIL: *
+// FIXME: Implement for other OSes and architectures.
+// REQUIRES: x86_64-target-arch, linux
```
Would you like to cover i386-target-arch as well?

chestnykh replied:
Done

chestnykh committed:
[compiler-rt] Add `DumpAllRegisters` impl for i386

- Restore dump_registers.cpp test in Posix/ dir
- Add linux-specifix tests inside Linux/ directory
Mac uses single dump_registers.cpp test inside Darwin/ directory
though checks in this test are partially 'restricted'
For linux we provide one test per each supported architecture with
checks for each part of `DumpAllRegisters` output

chestnykh commented:
@vitalybuka i've covered i386 arch and described in the last commit message why i restored common dump_register.cpp test and made two separate dump_register_<arch>.cpp tests instead

chestnykh committed:
[compiler-rt] Correct code style

vitalybuka committed:
Remove ALWAYS_INLINE
